XenModel/XenAPI-Extensions/VM.cs

```diff
@@ -774,13 +772,86 @@ public VirtualisationStatus GetVirtualisationStatus
                 }
                 else
                 {
-                    return VirtualisationStatus.OPTIMIZED;
+                    return VirtualisationStatus.IO_DRIVERS_INSTALLED | VirtualisationStatus.MANAGEMENT_INSTALLED;
+                }
+            }
+        }
+
+        private VirtualisationStatus GetVirtualisationStatusNewVM
+        {
+            get
+            {
+                Debug.Assert(HasNewVirtualisationStates);
+
+                var flags = HasRDP ? VirtualisationStatus.MANAGEMENT_INSTALLED : 0;
+
+                var vm_guest_metrics = Connection.Resolve(guest_metrics);
+                if (vm_guest_metrics != null && vm_guest_metrics.storage_paths_optimized && vm_guest_metrics.network_paths_optimized)
+                    flags |= VirtualisationStatus.IO_DRIVERS_INSTALLED;
+
+                if ((DateTime.UtcNow - BodgeStartupTime).TotalMinutes < 2)
+                {
+                    if (flags.HasFlag(VM.VirtualisationStatus.IO_DRIVERS_INSTALLED))
+                        return flags;
+
+                    return VirtualisationStatus.UNKNOWN;
+                }
+
+                return flags;
+            }
+        }
+
+        /// <summary>
+        /// Virtualization Status of the VM
+        /// </summary>
+        /// 
+        /// <remarks>
+        /// Following states are expected:
+        /// 
+        /// For Non-Windows VMs and for Windows VMs pre-Dundee:
+        ///   0 = Not installed
+        ///   1 = Unknown
+        ///   2 = Out of date
+        ///  12 = Tools installed (Optimized)
+        ///  
+        /// For Windows VMs on Dundee or higher:
+        ///    0 = Not installed
+        ///    1 = Unknown
+        ///    4 = I/O Optimized
+        ///   12 = I/O and Management installed
+        /// </remarks>
+        public VirtualisationStatus GetVirtualisationStatus

```

Comment:
I think this function is correct now. So we now just need to make sure we have the right state for all the usages.